joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/passes/TypeInferencePass.scala

```diff
@@ -87,19 +92,41 @@ class TypeInferencePass(cpg: Cpg) extends ConcurrentWriterCpgPass[Call](cpg) {
     cache.get(callKey).toScala.getOrElse {
       val callNameParts = getNameParts(call.name, call.methodFullName)
       resolvedMethodIndex.get(call.name).flatMap { candidateMethods =>
-        val candidateMethodsIter = candidateMethods.iterator
-        val uniqueMatchingMethod =
-          candidateMethodsIter.find(isMatchingMethod(_, call, callNameParts)).flatMap { method =>
-            val otherMatchingMethod = candidateMethodsIter.find(isMatchingMethod(_, call, callNameParts))
-            // Only return a resulting method if exactly one matching method is found.
-            Option.when(otherMatchingMethod.isEmpty)(method)
-          }
+        val uniqueMatchingMethod = retreiveMatchingMethod(candidateMethods, call, callNameParts) match {
+          case Some(method) => Some(method)
+          case None         => retreiveMatchingMethod(candidateMethods, call, callNameParts, ignoreArgTypes = true)
+        }
         cache.put(callKey, uniqueMatchingMethod)
         uniqueMatchingMethod
       }
     }
   }
 
+  /** Return a method only if there exists a one to one mapping of call to method node
+    * @param candidateMethods
+    * @param call
+    * @param callNameParts
+    * @param ignoreArgTypes
+    * @return
+    */
+  private def retreiveMatchingMethod(
+    candidateMethods: List[Method],
+    call: Call,
+    callNameParts: NameParts,
+    ignoreArgTypes: Boolean = false
+  ): Option[Method] = {
+    val candidateMethodsIter = candidateMethods.iterator
+    val uniqueMatchingMethod =
+      candidateMethodsIter.find(isMatchingMethod(_, call, callNameParts, ignoreArgTypes = ignoreArgTypes)).flatMap {
+        method =>
+          val otherMatchingMethod =
+            candidateMethodsIter.find(isMatchingMethod(_, call, callNameParts, ignoreArgTypes = ignoreArgTypes))
```

Comment:
An iterator can only be used once, and I see it used twice. Rather get rid of `candidateMethodsIter` and just call `candidateMethods.start` or `candidateMethods.iterator` when you need it.

Reply:
The reason for calling it twice is
- The first iterator call gives us the first match and the iterator stops there
- we use the same iterator to continue the search to see if we get another matching method

This is an optimization than calling the iterator from the start again

Reply:
No, iterators cannot be used more than once. I've verified this in a shell for you e.g.
```scala
scala> val x = Iterator(1, 2, 3)
val x: Iterator[Int] = <iterator>
                                                                                                                                                                                                                                                                                                                                                          
scala> x.toList
val res2: List[Int] = List(1, 2, 3)
                                                                                                                                                                                                                                                                                                                                                          
scala> x.toList
val res3: List[Int] = List()
```

Reply:
This is intentional (maybe a comment is required in code to explain this) and is an optimisation to avoid having to traverse the entire iterator twice.

The first call to `candidateMethodsIter.find` consumes iterator elements until the first match, but stops at this point. The second `find` call then continues the search until a second match is found.

To illustrate with an example considering only the first `find` call:
```
scala> val x = Iterator(1, 2, 3, 2, 4)
val x: Iterator[Int] = <iterator>
                                                                                               
scala> x.find(_ == 2)
val res3: Option[Int] = Some(2)
                                                                                               
scala> x.toList
val res4: List[Int] = List(3, 2, 4)
```
And for both calls:
```
scala> val x = Iterator(1, 2, 3, 2, 4)
val x: Iterator[Int] = <iterator>
                                                                                               
scala> x.find(_ == 2)
val res5: Option[Int] = Some(2)
                                                                                               
scala> x.find(_ == 2)
val res6: Option[Int] = Some(2)
                                                                                               
scala> x.toList
val res7: List[Int] = List(4)
```

Reply:
Ah now I understand, TDIL! Thanks